src/core/ddsi/src/ddsi_cdrstream.c

```diff
@@ -521,182 +608,464 @@ static const uint32_t *find_union_case (const uint32_t * __restrict union_ops, u
-    const void *valaddr = baseaddr + jeq_op[2];
+    void *valaddr = baseaddr + jeq_op[2];
     switch (valtype)
     {
-      case DDS_OP_VAL_1BY: dds_os_put1 (os, *(const uint8_t *) valaddr); break;
-      case DDS_OP_VAL_2BY: dds_os_put2 (os, *(const uint16_t *) valaddr); break;
-      case DDS_OP_VAL_4BY: dds_os_put4 (os, *(const uint32_t *) valaddr); break;
-      case DDS_OP_VAL_8BY: dds_os_put8 (os, *(const uint64_t *) valaddr); break;
-      case DDS_OP_VAL_STR: dds_stream_write_string (os, *(const char **) valaddr); break;
-      case DDS_OP_VAL_BST: dds_stream_write_string (os, (const char *) valaddr); break;
-      case DDS_OP_VAL_SEQ: case DDS_OP_VAL_ARR: case DDS_OP_VAL_UNI: case DDS_OP_VAL_STU:
-        dds_stream_write (os, valaddr, jeq_op + DDS_OP_ADR_JSR (jeq_op[0]));
+      case DDS_OP_VAL_1BY: *((uint8_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_2BY: *((uint16_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_4BY: case DDS_OP_VAL_ENU: *((uint32_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_8BY: *((uint64_t *) valaddr) = 0; break;
+      case DDS_OP_VAL_STR: *(char **) valaddr = dds_stream_reuse_string_empty (*((char **) valaddr)); break;
+      case DDS_OP_VAL_BST: case DDS_OP_VAL_BSP: case DDS_OP_VAL_SEQ: case DDS_OP_VAL_ARR: case DDS_OP_VAL_UNI: case DDS_OP_VAL_STU:
+        (void) dds_stream_skip_default (valaddr, jeq_op + DDS_OP_ADR_JSR (jeq_op[0]));
         break;
+      case DDS_OP_VAL_EXT: {
+        abort (); /* not supported */
+        break;
+      }
     }
   }
   return ops;
 }
 
-static void dds_stream_write (dds_ostream_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+static const uint32_t *dds_stream_write_delimited (dds_ostream_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+{
+  dds_os_put4 (os, 0xffffffff);
+  uint32_t offs = os->m_index;
+  ops = dds_stream_write (os, data, ops + 1);
+  *((uint32_t *) (os->m_buffer + offs - 4)) = os->m_index - offs;
+  return ops;
+}
+
+static const uint32_t *dds_stream_write_delimitedLE (dds_ostreamLE_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+{
+  dds_os_put4LE (os, 0xffffffff);
+  uint32_t offs = os->x.m_index;
+  ops = dds_stream_writeLE (os, data, ops + 1);
+  *((uint32_t *) (os->x.m_buffer + offs - 4)) = ddsrt_toLE4u (os->x.m_index - offs);
+  return ops;
+}
+
+static const uint32_t *dds_stream_write_delimitedBE (dds_ostreamBE_t * __restrict os, const char * __restrict data, const uint32_t * __restrict ops)
+{
+  dds_os_put4BE (os, 0xffffffff);
+  uint32_t offs = os->x.m_index;
+  ops = dds_stream_writeBE (os, data, ops + 1);
+  *((uint32_t *) (os->x.m_buffer + offs - 4)) = ddsrt_toBE4u (os->x.m_index - offs);
+  return ops;
+}
+
+static uint32_t get_length_code_seq (const uint32_t subtype)
+{
+  switch (subtype)
+  {
+    /* Sequence length can be used as byte length */
+    case DDS_OP_VAL_1BY:
+      return 5;
```

Comment:
Looks to me like these length codes could become #defines.

Reply:
Yes, that's definitely better, fixed in 1b3eb65